docs(workshops): add ESP RainMaker Studio workshop (Rainbow LED on ESP32-C3) (#756)
espressif/developer-portal: content/workshops/rainmaker-studio/_index.md

---

> title: "Build a Cloud-connected Custom Device with ESP RainMaker Studio"

Reviewer: At least at this point, it is not clear what `Custom Device` means. Why is it custom?

Maybe we don't need this word in the article title?

Author: I removed `custom` And just cloud-connected device.

date: "2026-06-18"

> lastmod: "2026-06-18"

Reviewer: I see that you added `lastmod` in the front matter. Normally, this field should only be added if you modify already published content, which is not your case. Actually, some other author also did it recently.

So I am curious what made you add `lastmod`? Maybe there are wrong instructions somewhere that I am not aware of? :)

Author: Hi, @f-hollow thank you for reviewing the document and for your feedback. I've made the requested changes. Please find the updated as follows. Thank you!

I think i just followed an existing template which had the `lastmod`

Author: removed lastmod from _index.md and from all three parts (Part 1, 2, 3).

tags: ["Workshop", "ESP RainMaker Studio", "ESP-IDF", "ESP32-C3", "IoT", "Cloud Connectivity"]
authors:
  - "ivan-theng"
showTableOfContents: false
---

Welcome to the ESP RainMaker Studio workshop. In this hands-on guide, you will go from a blank project to a cloud-connected Rainbow LED on an ESP32-C3 DevKit — with the RainMaker framework already wired up, so you can focus on your product instead of the boilerplate.

## What is ESP RainMaker Studio?

ESP RainMaker Studio is a browser-based drag-and-drop tool on the [ESP RainMaker Evaluation Hub](https://evaluation.rainmaker.espressif.com). It lets you visually design an ESP RainMaker **device data model** — the node, devices, and parameters that define what your IoT product exposes to the cloud and phone app — and then automatically generates a complete, buildable ESP-IDF project from that model.

If you are new to ESP RainMaker development, we recommend reviewing the [basic concepts and features of ESP RainMaker](https://docs.rainmaker.espressif.com/docs/product_overview/technical_overview/introduction/) before you begin.

> Skip the early setup work that usually slows down a RainMaker project: deciding how to structure the node and devices, mapping product controls into RainMaker parameters, wiring services such as provisioning and OTA, and learning enough of the framework before you can see your product take shape. With Studio, you start from the product experience instead — what the user should see in the phone app, what values should sync through the cloud, and what behavior the device should expose.

Reviewer: This paragraph can be made more clear. I also recommend splitting it into two paragraphs: (1) problem and (2) solution.

```suggestion
Before your product starts taking shape, you usually need to decide how to structure the node and devices, map product controls into RainMaker parameters, wire services such as provisioning and OTA, and grow a bit more comfortable with the framework.

With ESP RainMaker Studio, you can skip the early setup work that slows down a RainMaker project and start from the product experience instead — what the user should see in the phone app, what values should sync through the cloud, and what behavior the device should expose.
```

Author: Done — the content has been split into two paragraphs under a new "Skip Starting from Zero" subheading. The first paragraph highlights the challenge of setup overhead before a project can take shape, while the second introduces ESP RainMaker Studio as the solution.

[Source](https://github.com/ivanthengwr/developer-portal/blob/039662f5c76fbe139aebe9f9e8c1c6581bf8a4f3/content/workshops/rainmaker-studio/_index.md#L24)

> No account is required. Projects are saved in your browser. The generated project contains all the RainMaker framework setup, provisioning, OTA, scheduling, and services wired up and ready to go. Concentrate on what your product actually intends to do through the cloud and phone app, then fill in the hardware driver functions behind those controls.

Reviewer: ```suggestion
No account is required. Projects are saved in your browser. The generated project contains all the RainMaker framework setup, provisioning, OTA, scheduling, and services wired up and ready to go. Focus on what your product actually intends to do through the cloud and phone app, then fill in the hardware driver functions behind those controls.
```

Author: This have been updated according to the above.

[Source](https://github.com/ivanthengwr/developer-portal/blob/039662f5c76fbe139aebe9f9e8c1c6581bf8a4f3/content/workshops/rainmaker-studio/_index.md?plain=1#L30)

> {{< figure

Reviewer: I recommend that you move this figure above the previous paragraph `No account is required...`

Because you introduce it in the paragraph just above that place.

Author: Done — moved the Studio landing page figure to the top of the "What is ESP RainMaker Studio?" section, so it appears where it is first introduced.

>   default=true
>   src="assets/01-studio-landing.webp"
>   caption="ESP RainMaker Studio on the Evaluation Hub homepage"
> >}}

Reviewer: Please remove `default=true` **here and in all other** `figure` shortcodes in your workshop. We previously used it as a workaround for some issue that has been already fixed.

```suggestion
{{< figure
  src="assets/01-studio-landing.webp"
  caption="ESP RainMaker Studio on the Evaluation Hub homepage"
>}}
```

Author: Done — removed default=true from all `{{< figure >}}` shortcodes across `_index.md` and all three parts.

## Why Use ESP RainMaker Studio?

> | Benefit | Detail |

Reviewer: ```suggestion
| Benefit | Details |
```

Author: Done - Typo Corrected

|---|---|
| **Complete ESP RainMaker Framework Setup** | `app_main.c` is fully generated — RainMaker init, NVS, network, OTA, timezone, scheduling, scenes, and Insights are all wired up |
| **Visual data model** | Design your device hierarchy graphically; see it reflected instantly in the JSON and Code tabs |

> | **Instant feedback** | The issues counter flags missing model/type fields before you can download |

Reviewer: > The issues counter flags missing model/type fields before you can download

Not clear where the verb is or how this sentence is structured. Maybe commas are missing?

Author: Done — reworded to: "Get instant feedback on missing fields such as model and type, so your codebase is correct and accurate before you download the     complete ESP-IDF project.

| **Chip-targeted output** | Select your target (ESP32-C3, ESP32-S3, etc.) and the project ships with the matching `sdkconfig.defaults` |
| **Prototype to product** | The same data model you define here maps directly to your production RainMaker configuration — no rework |
| **Faster time-to-market** | Skip days of reading RainMaker API docs for standard device types; focus engineering time on your hardware differentiation |
| **Standard param types** | Power, Brightness, Color Hue, Fan Speed, and more are pre-configured with the correct ESP RainMaker type strings, UI hints, and default values |
| **Custom params supported** | Add any custom parameter type with a slider, toggle, or input UI — full flexibility for novel devices |

## Prerequisites

| Requirement | Details |
|---|---|

> | **Browser** | Chrome or Edge (Chromium-based). Firefox and Safari are not supported for flashing via Web Serial |

Reviewer: I recommend adding a line break to visually separate supported and non-supported browsers.

```suggestion
| **Browser** | Chrome or Edge (Chromium-based).<br>Firefox and Safari are not supported for flashing via Web Serial |
```

Author: Done — added <br> to visually separate the supported and unsupported browser lines in the Prerequisites table.

| **ESP-IDF** | v5.0 or later. Install via the [ESP-IDF Getting Started Guide](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/get-started/) |
| **Hardware** | ESP32-C3-DevKitC. The onboard WS2812 RGB LED is on GPIO 8, BOOT button on GPIO 9 |
| **ESP RainMaker Phone App** | [iOS](https://apps.apple.com/us/app/esp-rainmaker-home/id1563728960) or [Android](https://play.google.com/store/apps/details?id=com.espressif.novahome&hl) — needed to provision and control the device |

{{< alert icon="circle-info" cardColor="#b3e0f2" iconColor="#04a5e5">}}

> The Studio runs entirely in your browser. No ESP-IDF setup is needed to design the data model — only for building and flashing the downloaded project.

Reviewer: ```suggestion
The Studio runs entirely in your browser. No ESP-IDF setup is needed to design the data model. It is only needed later for building and flashing the downloaded project.
```

Author: Done - Applied this change

{{< /alert >}}

## Agenda

This workshop is divided into three parts:

- **[Part 1 — Build Your Data Model in Studio](part-1/)**: Open Studio, create a project, add a custom Rainbow LED device, configure its parameters, and download the generated project.
- **[Part 2 — Implement the Driver Functions](part-2/)**: Understand the generated scaffold and fill in the hardware driver — LED strip initialisation, button handling, and the FreeRTOS rainbow task.
- **[Part 3 — Build, Flash, and Test](part-3/)**: Build the project with ESP-IDF, flash it to your ESP32-C3, provision it with the RainMaker phone app, and verify every control.

{{< alert icon="mug-hot" >}}
**Estimated time: 90 min**
{{< /alert >}}

## Next Step

> The next step is **[Part 1 — Build Your Data Model in Studio](part-1/)**.

## Feedback

> If you have any feedback about this workshop, feel free to start a new [discussion on GitHub](https://github.com/espressif/developer-portal/discussions).

Reviewer: @pedrominatel Regarding the possible proposals on how to improve this workshops, should discussions be created in the Developer Portal repo?

It seems like this kind of discussions should be help in a more appropriate place like RainMaker documentation repo, etc.,, where more people related to this project can naturally participate.

What do you think?

Reviewer 2: @f-hollow and @ivanthengwr

My recommendation is to use the feedback form for suggestions and improvements, and the RainMaker GitHub repo for questions and issues.
The feedback form MR was created and should be merged in the following days. After merging, all articles and pages will include the icon for feedback.

For now, we can remove the feedback section.

Reviewer: @ivanthengwr Please remove the feedback section. As Pedro mentioned, we will add a feedback form for the whole developer portal in the nearest future. 

---

~~## Feedback~~

~~If you have any feedback about this workshop, feel free to start a new [discussion on GitHub](https://github.com/espressif/developer-portal/discussions).~~

Author: Hi @f-hollow @pedrominatel , this feedback section have been removed in the latest commit

---
title: "Build a Cloud-connected device with ESP RainMaker Studio"

> date: "2026-07-08"

Reviewer: @ivanthengwr If you need to have it merged today, please change the date here and in all parts to 9 July.

Author: Got it, I have change the date to 9 July

> summary: "Learn to design an ESP RainMaker device data model using the Studio tool on Evaluation Hub, implement the ESP32-C3 hardware driver, and build, flash, and control a cloud-connected Rainbow LED from the ESP RainMaker Home app."

Reviewer: ```suggestion
summary: "Learn to design an ESP RainMaker device data model using the Studio tool on Evaluation Hub, implement the ESP32-C3 hardware driver, and build, flash, and control a cloud-connected Rainbow LED from the ESP RainMaker Home app."
featureAsset: "img/featured/rainmaker-workshop-background.webp"
```

Author: This addition have been added to all parts.

tags: ["Workshop", "ESP RainMaker Studio", "ESP-IDF", "ESP32-C3", "IoT", "Cloud Connectivity"]
authors:
  - "ivan-theng"
showTableOfContents: false
---

> Welcome to the ESP RainMaker Studio workshop. In this hands-on guide, you will go from a blank project to a cloud-connected Rainbow LED on an ESP32-C3 DevKit — with the RainMaker framework already wired up, so you can focus on your product instead of the boilerplate.

Reviewer 3: Could you remove the em dash? The problem with that is that it shouts LLM :D

Author: Thank you for the feedback @FBEZ! Please find my response below

Author: I have also asked the cursor agent to remove all em dash as necessary and reviewed it. 😄

## What is ESP RainMaker Studio?

{{< figure
  src="assets/01-studio-landing.webp"
  caption="ESP RainMaker Studio on the Evaluation Hub homepage"
>}}

> ESP RainMaker Studio is a browser-based drag-and-drop tool on the [ESP RainMaker Evaluation Hub](https://evaluation.rainmaker.espressif.com). It lets you visually design an ESP RainMaker **device data model** — the node, devices, and parameters that define what your IoT product exposes to the cloud and phone app — and then automatically generates a complete, buildable ESP-IDF project from that model.

Reviewer 3: Here the same. I'd split the sentences, since the definition of datamodel is quite long.
```suggestion
ESP RainMaker Studio is a browser-based drag-and-drop tool on the [ESP RainMaker Evaluation Hub](https://evaluation.rainmaker.espressif.com). It lets you visually design an ESP RainMaker **device data model**, which is the node, devices, and parameters that define what your IoT product exposes to the cloud and phone app. From that model, ESP RainMaker Studio automatically generates a complete, buildable ESP-IDF project.
```

Author: This change have been applied

If you are new to ESP RainMaker development, we recommend reviewing the [basic concepts and features of ESP RainMaker](https://docs.rainmaker.espressif.com/docs/product_overview/technical_overview/introduction/) before you begin.

> ### Skip Starting from Zero

Reviewer 3: This sounds like a something to do instead of an advantage of why using ESP Rainmaker. I'd remove this section and put a shorter version of it into the next section.

Author: Yes, indeed. I've moved it to the beginning of the workshop introduction. The intention is to encourage developers and readers who are new to RainMaker to review the introductory RainMaker documentation before proceeding. This helps ensure they are familiar with the RainMaker-specific terminology used throughout the workshop and won't get lost along the way.

Before your product starts taking shape, you usually need to decide how to structure the node and devices, map product controls into RainMaker parameters, wire services such as provisioning and OTA, and grow a bit more comfortable with the framework. That takes time and effort — often starting with reading the documentation and experimenting before anything comes together.

> With ESP RainMaker Studio, you don't have to start from zero. Skip the early setup that slows down a project kickoff and design from the product experience first — what users see in the phone app, what syncs through the cloud, and what the device should do.

Reviewer 3: Same here, al later on.

Author: I've reformatted it into clear bullet points to highlight the key benefits and what readers can expect to gain from using Studio.

| **Chip-targeted output** | Select your target (ESP32-C3, ESP32-S3, etc.) and the project ships with the matching `sdkconfig.defaults` |
| **Prototype to product** | The same data model you define here maps directly to your production RainMaker configuration — no rework |
| **Faster time-to-market** | Skip days of reading RainMaker API docs for standard device types; focus engineering time on your hardware differentiation |
| **Standard param types** | Power, Brightness, Color Hue, Fan Speed, and more are pre-configured with the correct ESP RainMaker type strings, UI hints, and default values |
| **Custom params supported** | Add any custom parameter type with a slider, toggle, or input UI — full flexibility for novel devices |

## Prerequisites

| Requirement | Details |
|---|---|
| **Browser** | Chrome or Edge (Chromium-based). <br> Firefox and Safari are not supported for flashing via Web Serial |
| **ESP-IDF** | v5.0 or later. Install via the [ESP-IDF Getting Started Guide](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/get-started/) |
| **Hardware** | ESP32-C3-DevKitC. The onboard WS2812 RGB LED is on GPIO 8, BOOT button on GPIO 9 |

> | **ESP RainMaker Phone App** | [iOS](https://apps.apple.com/us/app/esp-rainmaker-home/id1563728960) or [Android](https://play.google.com/store/apps/details?id=com.espressif.novahome&hl) — needed to provision and control the device |

Reviewer: Is the iOS link correct?

When I click it, I somehow get redirected to `https://apps.apple.com/cn/iphone/today`.

Author: I have change the link to a common reference point in RainMaker docs where the latest iOS and android download links for RainMaker Home App can be found.
